### PR TITLE
Implement "gap" property for flexbox algorithm

### LINKED ISSUE #66: Implement "gap" property (part of the spec for flex layout)
## What problem does this solve or what need does it fill?

The issue is as old as flex box: spacing between elements

## What solution would you like?

Implement the spec: https://developer.mozilla.org/en-US/docs/Web/CSS/gap#specifications
Or at least a reasonable subset, note that `calc` should be probably outside the scope

Thus, I think, implementing just explicit values should be good enough (at least as a start)

from spec:
```css
gap: 20px; /* <--- that would be totally enough for me */


/* these are probably way less useful and applicable */
gap: 1em;
gap: 3vmin; 
gap: 0.5cm;
/* One <percentage> value */
gap: 16%;
gap: 100%;

```

## What alternative(s) have you considered?

Work around is simple: margins + negative padding, but super cumbersome

### PR DESCRIPTION
**Pull request by nicoburns**
# Objective

- Implements flexbox "gap" property (fixes #66)
- Adds support for CSS grid properties to the test generation script (I had already added gap as part of this work, so it made sense to pull that in for this - it doesn't actually effect the generated tests because the fixtures don't actually include any grid properties).

## Context

Yoga have recently implemented this. I pulled in their tests. They don't support percentage gaps, so we may need some more testing around that (or to remove that feature from this implementation).

This builds upon #246, which should be merged first.

### CODE REVIEW

**Comment by nicoburns** [comment]
Update: percentage gaps work fine if the parent node has a defined size, but they don't match Chrome/Firefox if the parent does not have a defined size.

**Comment by nicoburns** [comment]
Ok, I think I have percentage gaps with undefined container sizes working properly (matching Chrome/Firefox) now. I have to admit that this was a bit of a trial and error process, and I don't fully understand why it works the way it does. But I'm fairly to happy to merge it anyway as:

- I have generated tests that cover a few different cases
- The changes were relatively small, and they don't regress any of the existing tests.
- I can't imagine anyone actually wanting to use percentage gaps with undefined container sizes: the result is that the children overflow the container. I imagine that the vast majority of people will use fixed gaps AND also defined container sizes (i.e. they will avoid this code path in two separate ways).

This PR is now ready for review (although in practice is might be best to wait until to #246 is merged in order to get a better diff).

**Comment by alice-i-cecile** [comment]
#246 is merged! Can you rebase so I can review without losing my mind?

**Comment by nicoburns** [comment]
@alice-i-cecile rebased :)

File: src/compute/flexbox.rs [hunk 1]
```diff
@@ -621,17 +637,24 @@ fn collect_flex_lines<'a>(
...
+        let main_axis_gap = constants.gap.main(constants.dir);
 
         while !flex_items.is_empty() {
             let mut line_length = 0.0;
             let index = flex_items
                 .iter()
                 .enumerate()
                 .find(|&(idx, child)| {
-                    line_length += child.hypothetical_outer_size.main(constants.dir);
+                    let gap_contribution = if idx == 0 { 0.0 } else { main_axis_gap };
```
Comment (alice-i-cecile) [inline src/compute/flexbox.rs:656]:
```suggestion
                    // The first item in a column / row does not have a gap
                    let gap_contribution = if idx == 0 { 0.0 } else { main_axis_gap };
```

File: src/compute/flexbox.rs [hunk 2]
```diff
@@ -660,12 +683,16 @@ fn resolve_flexible_lengths(
     constants: &AlgoConstants,
     available_space: Size<AvailableSpace>,
 ) {
+    let total_main_axis_gap =
```
Comment (alice-i-cecile) [inline src/compute/flexbox.rs:686]:
```suggestion
   // Lines with only one item don't care about gaps
    let total_main_axis_gap =
```

File: src/compute/flexbox.rs [hunk 3]
```diff
@@ -1161,7 +1197,10 @@ fn distribute_remaining_free_space(
     constants: &AlgoConstants,
 ) {
     for line in flex_lines {
-        let used_space: f32 = line.items.iter().map(|child| child.outer_target_size.main(constants.dir)).sum();
+        let total_main_axis_gap =
+            if line.items.len() > 1 { constants.gap.main(constants.dir) * (line.items.len() - 1) as f32 } else { 0.0 };
```
Comment (alice-i-cecile) [inline src/compute/flexbox.rs:1201]:
```suggestion
           // Lines with only one item don't care about gaps
            if line.items.len() > 1 { constants.gap.main(constants.dir) * (line.items.len() - 1) as f32 } else { 0.0 };
```

File: src/compute/flexbox.rs [hunk 4]
```diff
@@ -1388,21 +1442,23 @@ fn determine_container_cross_size(
     node_size: Size<Option<f32>>,
     constants: &mut AlgoConstants,
 ) -> f32 {
-    let total_cross_size: f32 = flex_lines.iter().map(|line| line.cross_size).sum();
+    let gap_contribution =
```
Comment (alice-i-cecile) [inline src/compute/flexbox.rs:1445]:
```suggestion
    // Lines with only one item don't care about gaps
    let gap_contribution =
```

File: src/compute/flexbox.rs [hunk 5]
```diff
@@ -1418,48 +1474,70 @@ fn align_flex_lines_per_align_content(
     constants: &AlgoConstants,
     total_cross_size: f32,
 ) {
-    let free_space = constants.inner_container_size.cross(constants.dir) - total_cross_size;
     let num_lines = flex_lines.len();
+    let gap = constants.gap.cross(constants.dir);
+    let total_cross_axis_gap = if num_lines > 1 { gap * (num_lines - 1) as f32 } else { 0.0 };
```
Comment (alice-i-cecile) [inline src/compute/flexbox.rs:1479]:
```suggestion
   // // Lines with only one item don't care about gaps
    let total_cross_axis_gap = if num_lines > 1 { gap * (num_lines - 1) as f32 } else { 0.0 };
```

**Comment by alice-i-cecile** [comment]
I'm very happy with this. It's great to see how far the code base has come. And thank goodness for the generated tests.

I've suggested some comments to help new readers get more comfortable with the "why" of gaps.

**Comment by alice-i-cecile** [comment]
Can we add an example demonstrating this?

File: scripts/gentest/src/main.rs [hunk 6]
```diff
@@ -203,18 +210,95 @@ fn generate_assertions(ident: &str, node: &json::JsonValue) -> TokenStream {
     )
 }
 
+// type QuoteProp = Fn(prop_name: &str, quoter: impl Fn(&json::object::Object) -> TokenStream) -> TokenStream;
```
Comment (Weibye) [inline scripts/gentest/src/main.rs:213]: Is this something we like to keep around? If yes, please add a comment explaining why, if no then remove it?

File: src/style.rs [hunk 7]
```diff
@@ -433,6 +436,7 @@ impl FlexboxLayout {
         margin: Rect::UNDEFINED,
         padding: Rect::UNDEFINED,
         border: Rect::UNDEFINED,
+        gap: Size { width: Dimension::Points(0.0), height: Dimension::Points(0.0) },
```
Comment (Weibye) [inline src/style.rs:439]: Should the default gap be `Size::UNDEFINED`?

**Comment by Weibye** [comment]
Also: ❤️ @nicoburns You've put in amazing work here over the last few weeks / days. Really appreciate it!

File: scripts/gentest/src/main.rs (hunk 6 above)
Comment (alice-i-cecile) [reply scripts/gentest/src/main.rs:213]:
```suggestion
```

File: src/style.rs (hunk 7 above)
Comment (nicoburns) [reply src/style.rs:439]: Hmm... yes, I suppose it probably should, although it doesn't make much difference as `UNDEFINED` is going go resolve to 0 anyway. It'll make the code look nicer though :)

**Comment by alice-i-cecile** [comment]
Not merging until that regression is fixed, but the code itself is in good shape.

**Comment by nicoburns** [comment]
Regression fixed :)

**Pull request merged**